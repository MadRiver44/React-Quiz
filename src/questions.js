// const questions = [
//   {
//     question: 'What is Redux?',
//     answers: [
//       'A state management tool',
//       'A tablet for indigestion',
//       'The newest version of JavaScipt',
//       'Nessecary for a React project',
//     ],
//     answerScore: '',
//   },
//   {
//     question: 'React was developed by:',
//     answers: [
//       'U.S government',
//       "React is a conspiracy, doesn't really exist",
//       'Facebook',
//       'Google',
//     ],
//   },
//   {
//     question: 'What is a pure function?',
//     answers: [
//       'A generic resusable function',
//       'A function that does not mutate its arguments nor produces side effects',
//       'A generator',
//       'Returns a long integer',
//     ],
//   },
//   {
//     question: 'A container component is:',
//     answers: [
//       'Handles the business logic and state and passes props to a presentational component',
//       'Same as a Presentational component',
//       'Invoked only once per react app',
//       'A node module',
//     ],
//   },
//   {
//     question: 'What does webpack do?',
//     answers: [
//       'Compiles sass and css files',
//       'Bundles all modules into a single file and performs optimizations',
//       'Packs an html and css file together',
//       'Creates a server for json files',
//     ],
//   },
// ];

const questions = [
  {
    question: 'What is Redux?',
    answers: [
      {
        answer: 'A state management tool',
        points: 3,
      },
      {
        answer: 'A tablet for indigestion',
        points: 0,
      },
      {
        answer: 'The newest version of JavaScipt',
        points: 2,
      },
      {
        answer: 'Nessecary for a React project',
        points: 1,
      },
    ],
    correctAnswer: 'answer1',
  },
  {
    question: 'React was developed by:',
    answers: [
      {
        answer: 'U.S government',
        points: 0,
      },
      {
        answer: "React is a conspiracy, doesn't really exist",
        points: 1,
      },
      {
        answer: 'Facebook',
        points: 3,
      },
      {
        answer: 'Google',
        points: 2,
      },
    ],
    correctAnswer: 'answer3',
  },
  {
    question: 'What is a pure function?',
    answers: [
      {
        answer: 'A generic resusable function',
        points: 2,
      },
      {
        answer: 'A function that does not mutate its arguments nor produces side effects',
        points: 3,
      },
      {
        answer: 'A generator',
        points: 1,
      },
      {
        answer: 'Returns a long integer',
        points: 0,
      },
    ],
    correctAnswer: 'answer2',
  },
  {
    question: 'A container component is:',
    answers: [
      {
        answer:
          'Handles the business logic and state and passes props to a presentational component',
        points: 3,
      },
      {
        answer: 'Same as a Presentational component',
        points: 2,
      },
      {
        answer: 'Invoked only once per react app',
        points: 1,
      },
      {
        answer: 'A node module',
        points: 0,
      },
    ],
    correctAnswer: 'answer1',
  },
  {
    question: 'What does webpack do?',
    answers: [
      {
        answer: 'Compiles sass and css files',
        points: 2,
      },
      {
        answer: 'Bundles all modules into a single file and performs optimizations',
        points: 3,
      },
      {
        answer: 'Packs an html and css file together',
        points: 1,
      },
      {
        answer: 'Creates a server for json files',
        points: 0,
      },
    ],
  },
];

export default questions;
