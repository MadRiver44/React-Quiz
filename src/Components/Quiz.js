import React, { Component } from 'react';
import questions from '../questions.js';
import Question from './Question';
import Answers from './Answers';
import Assesment from './Assesment';

const answers = questions.map((q, index) => q.answers); // array of array of answer objects
// const results = answers[0].map((a, index) => a.answer); // array of answers
// console.log(answers, results);
const results = (function a(answers) {
  let arr = [];
  for (let i = 0; i < answers.length; i++) {
    let item = Array.from(answers[i].map((item, index) => item.answer));
    if (item.length !== 0) {
      arr.push(item);
    }
  }
  return arr;
})(answers);

const calculateScore = str => {
  let value = 0;
  for (let i = 0; i < answers.length; i++) {
    let obj = answers[i];
    for (let key in obj) {
      if (obj[key]['answer'] === str) {
        value = obj[key]['points'];
        return value;
      }
    }
  }
};

class Quiz extends Component {
  constructor(props) {
    super(props);
    this.state = {
      questions: [...questions],
      answers: [...results], // needs to be an array of arrays of answers
      correctAnswerCount: 0,
      almostRightCount: 0,
      wayOffCount: 0,
      deadWrongCount: 0,
      score: [],
      scoreTally: 0,
      count: 0,
      missedQuestion: false,
    };
    this.handleClick = this.handleClick.bind(this);
    this.questionTypeCount = this.questionTypeCount.bind(this);
    this.handleChangeAnswerLeft = this.handleChangeAnswerLeft.bind(this);
    this.handleChangeAnswerRight = this.handleChangeAnswerRight.bind(this);
    // this.detectUnansweredQuestions = this.detectUnansweredQuestions.bind(this);
  }

  // detectUnansweredQuestions() {
  //   let indexesOfSkippedQuestions = this.state.score.reduce((all, item, index) => {
  //     if (item === null) {
  //       all.push(item);
  //     }
  //     return all;
  //   }, []);
  //   if (indexesOfSkippedQuestions.length !== 0) {
  //     this.setState({
  //       unansweredQuestions: indexesOfSkippedQuestions,
  //       missedQuestion: !!this.state.missedQuestion,
  //     });
  //   }
  //   console.log(indexesOfSkippedQuestions);
  // }

  handleChangeAnswerLeft(e) {
    e.preventDefault();
    this.setState({
      count: this.state.count - 1,
      missedQuestion: !this.state.missedQuestion,
    });
  }

  handleChangeAnswerRight(e) {
    // if i go back one, answer it and then click right again,
    // it looks as if I still didn't answer the last question
    e.preventDefault();
    this.setState({
      count: this.state.count + 1,
    });

    if (this.state.score[this.state.count - 1] !== null) {
      this.setState({
        missedQuestion: !this.state.missedQuestion,
      });
    }
  }

  questionTypeCount() {
    let arr = [...this.state.score];
    let deadWrong, wayOff, almostRight, correct;
    deadWrong = arr.filter(item => item === 0).length;
    wayOff = arr.filter(item => item === 1).length;
    almostRight = arr.filter(item => item === 2).length;
    correct = arr.filter(item => item === 3).length;
    this.setState({
      deadWrongCount: deadWrong,
      wayOffCount: wayOff,
      almostRightCount: almostRight,
      correctAnswerCount: correct,
    });
  }

  handleClick(e) {
    e.preventDefault();
    let points = calculateScore(e.target.value);
    let arrayOfScores = this.state.score;
    arrayOfScores[this.state.count] = points;
    let total = arrayOfScores.reduce((item, itemNext) => {
      return item + itemNext;
    }, 0);
    this.setState({
      count: this.state.count + 1,
      score: arrayOfScores, //this.state.score + points
      scoreTally: total,
    });
    if (this.state.score.length) {
      this.questionTypeCount();
    }
  }

  render() {
    const {
      questions,
      answers,
      count,
      correctAnswerCount,
      score,
      almostRightCount,
      wayOffCount,
      deadWrongCount,
      scoreTally,
      missedQuestion,
    } = this.state;

    return count !== results.length ? (
      <div className="question">
        <span className="questionNumber">
          {count + 1} {count + 1 === 5 ? <span>Last question</span> : null}
        </span>

        <Question answer={answers[count]} count={count}>
          {questions[count].question}
          <span className="alert">
            {missedQuestion === true ? <p>Hey you misssed last question: # {count}</p> : null}
          </span>
        </Question>

        <div>
          <span style={{ fontSize: '20px' }}>{'Questions Answered : '}</span>
          <span className="alert-green">
            {`${this.state.score.map((item, index) => index + 1)}`}{' '}
          </span>
        </div>

        <Answers handleClick={this.handleClick} answer={answers[count]} count={count} />

        <div className="arrows">
          {count === 0 ? (
            <button className="hidden btn btn-left" />
          ) : (
            <button className="btn btn-left" onClick={this.handleChangeAnswerLeft}>
              {'<-'}
            </button>
          )}
          <button className="btn btn-right" onClick={this.handleChangeAnswerRight}>
            {'->'}
          </button>
        </div>
      </div>
    ) : (
      <Assesment
        count={count}
        correctAnswerCount={correctAnswerCount}
        almostRightCount={almostRightCount}
        wayOffCount={wayOffCount}
        deadWrongCount={deadWrongCount}
        scoreTally={scoreTally}
      />
    );
  }
}
export default Quiz;
