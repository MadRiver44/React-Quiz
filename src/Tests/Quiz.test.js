import React from 'react';
import renderer from 'react-test-renderer';
import Quiz from '../Components/Quiz';
import questions from './../questions';

describe('Quiz component', () => {
  test('Quiz renders a question and an array of answers', () => {
    const tree = renderer
      .create(
        <Quiz
          question={'What is Redux?'}
          answer={[
            'A state management tool',
            'A tablet for indeigestion',
            'The newest version of JavaScipt',
            'Nessecary for a React project',
          ]}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
