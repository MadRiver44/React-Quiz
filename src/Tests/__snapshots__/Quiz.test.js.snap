// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz component Quiz renders a question and an array of answers 1`] = `
<div
  className="question"
>
  <div>
    <h1>
      What is Redux?
    </h1>
  </div>
  <div>
    <form>
      <ul>
        <li>
          <button
            className="answers btn"
            onClick={[Function]}
            value="A state management tool"
          >
            A state management tool
          </button>
        </li>
        <li>
          <button
            className="answers btn"
            onClick={[Function]}
            value="A tablet for indigestion"
          >
            A tablet for indigestion
          </button>
        </li>
        <li>
          <button
            className="answers btn"
            onClick={[Function]}
            value="The newest version of JavaScipt"
          >
            The newest version of JavaScipt
          </button>
        </li>
        <li>
          <button
            className="answers btn"
            onClick={[Function]}
            value="Nessecary for a React project"
          >
            Nessecary for a React project
          </button>
        </li>
      </ul>
    </form>
  </div>
</div>
`;
